### PR TITLE
Set max_score to NaN in empty SearchHits rather than 0

### PR DESCRIPTION
**Pull request by javanna**
In some cases, for instance when we have no shards to search on, we
return an empty instance of SearchHits. We may or may not be sorting on
score, yet when we have no results the max_score should be set to NaN
(which converts to null on the xContent representation), while we
currently set it to `0`.

### CODE REVIEW

**Review by cbuescher — APPROVED** [review]
LGTM, makes more sense to me intuitively.

**Comment by javanna** [comment]
@jimczi do you have thoughts on this change? I marked this breaking just to make sure, and I plan to merge it to master only.

**Review by jpountz — APPROVED** [review]
+1 Maybe the SearchHits constructor should fail if an empty hits array is passed and maxScore is not NaN?

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 1]
```diff
@@ -383,7 +383,8 @@ private SearchHits getHits(ReducedQueryPhase reducedQueryPhase, boolean ignoreFr
             }
         }
         return new SearchHits(hits.toArray(new SearchHit[0]), reducedQueryPhase.totalHits,
-            reducedQueryPhase.maxScore, sortedTopDocs.sortFields, sortedTopDocs.collapseField, sortedTopDocs.collapseValues);
+            hits.isEmpty() ? Float.NaN : reducedQueryPhase.maxScore,
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: I am not sure that this is right, we should rather track the root cause for this. It gets quite complicated though as there are places where we do have empty hits and maxScore set to some value. I've seen quite some scroll tests fail due to this, as well as others.

File: server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java [hunk 2]
```diff
@@ -181,7 +181,8 @@ public void execute(SearchContext context) {
             }
 
             TotalHits totalHits = context.queryResult().getTotalHits();
-            context.fetchResult().hits(new SearchHits(hits, totalHits, context.queryResult().getMaxScore()));
+            context.fetchResult().hits(new SearchHits(hits, totalHits, hits.length == 0 ? Float.NaN : context.queryResult().getMaxScore()));
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java]: here too, not sure this is right. feels like cheating to me :)

**Comment by javanna** [comment]
@jpountz I tried to enforce that maxScore is set to NaN whenever we have empty hits, but that is not as easy as it initially sounded. I left comments on two changes that I had to make, which are not great in my opinion. Also, I still have test failures that I need to track down. I am thinking that if we want to pursue the enforcing idea, we should do it as a followup and understand better whether this makes sense, also think about the bw comp aspects.

**Comment by jimczi** [comment]
This is outdated, hence closing.

**Comment by javanna** [comment]
It is true that it is outdated, yet it may still be something that we want to fix. I remember finding it harder than expected and then leaving it hanging for lack of feedback. I can pick it back up if we think it's useful.

**Pull request closed without merging**